cpp/src/arrow/tensor.cc

```diff
@@ -123,6 +123,80 @@ Type::type Tensor::type_id() const { return type_->id(); }
 
 bool Tensor::Equals(const Tensor& other) const { return TensorEquals(*this, other); }
 
+namespace {
+
+template <typename TYPE>
+size_t StridedTensorCountNonZero(int dim_index, int64_t offset, const Tensor& tensor) {
+  using c_type = typename TYPE::c_type;
+  c_type const zero = c_type(0);
+  size_t nnz = 0;
+  if (dim_index == tensor.ndim() - 1) {
+    for (int64_t i = 0; i < tensor.shape()[dim_index]; ++i) {
+      auto const* ptr = tensor.raw_data() + offset + i * tensor.strides()[dim_index];
+      auto& elem = *reinterpret_cast<c_type const*>(ptr);
+      if (elem != zero) ++nnz;
+    }
+    return nnz;
+  }
+  for (int64_t i = 0; i < tensor.shape()[dim_index]; ++i) {
+    nnz += StridedTensorCountNonZero<TYPE>(dim_index + 1, offset, tensor);
+    offset += tensor.strides()[dim_index];
+  }
+  return nnz;
+}
+
+template <typename TYPE>
+size_t ContiguousTensorCountNonZero(const Tensor& tensor) {
+  using c_type = typename TYPE::c_type;
+  auto* data = reinterpret_cast<c_type const*>(tensor.raw_data());
+  return std::count_if(data, data + tensor.size(),
+                       [](c_type const& x) { return x != 0; });
+}
+
+template <typename TYPE>
+inline size_t TensorCountNonZero(const Tensor& tensor) {
+  if (tensor.is_contiguous()) {
+    return ContiguousTensorCountNonZero<TYPE>(tensor);
+  } else {
+    return StridedTensorCountNonZero<TYPE>(0, 0, tensor);
+  }
+}
+
+}  // namespace
+
+size_t Tensor::CountNonZero() const {
+  if (size() == 0) {
+    return 0;
+  }
+
+  switch (type()->id()) {
+    case Type::UINT8:
+      return TensorCountNonZero<UInt8Type>(*this);
+    case Type::INT8:
+      return TensorCountNonZero<Int8Type>(*this);
+    case Type::UINT16:
+      return TensorCountNonZero<UInt16Type>(*this);
+    case Type::INT16:
+      return TensorCountNonZero<Int16Type>(*this);
+    case Type::UINT32:
+      return TensorCountNonZero<UInt32Type>(*this);
+    case Type::INT32:
+      return TensorCountNonZero<Int32Type>(*this);
+    case Type::UINT64:
+      return TensorCountNonZero<UInt64Type>(*this);
+    case Type::INT64:
+      return TensorCountNonZero<Int64Type>(*this);
+    case Type::HALF_FLOAT:
+      return TensorCountNonZero<HalfFloatType>(*this);
+    case Type::FLOAT:
+      return TensorCountNonZero<FloatType>(*this);
+    case Type::DOUBLE:
+      return TensorCountNonZero<DoubleType>(*this);
+    default:
+      return 0;  // This shouldn't be unreachable
```

Comment:
I could theoretically define a tensor of timestamps, bools or decimals. I'm not sure returning 0 is the right answer here. Perhaps the method should return a `Status` instead?

Reply:
I agree with you about returning zero is wrong.  It, however, is prevented to create tensor with Timestamp type due to checking by `is_tensor_supported` in the constructor.  I changed the return value to `-1`, tentatively.

What we should guarantee here is we can notice that we have to change `CountNonZero` when someone adds new tensor-supported types.  I'm considering the way to do so.

Reply:
We may not need a want to count non-zeros for some tensor types, for example dates and times. So I think returning a `Status` may be the best thing here.